core/node/rpc/node2nodeauth/cert.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	sigPublicKey, err := ethcrypto.SigToPub(
+		crypto.TownsHashForCert.Hash(publicKeyDER).Bytes(),
+		certExt.Signature,
+	)
+	if err != nil {
+		return AsRiverError(err, Err_UNAUTHENTICATED).Message("Failed to extract public key from the cert signature").LogError(logger)
+	}
+	recoveredAddr := ethcrypto.PubkeyToAddress(*sigPublicKey)
+
+	if recoveredAddr.Cmp(common.HexToAddress(certExt.Address)) != 0 {
+		return RiverError(Err_UNAUTHENTICATED, "Node-2-node cert signature does not match address").LogError(logger)
+	}
+
+	node, err := nodeRegistry.GetNode(recoveredAddr)
+	if err != nil {
+		return err
+	}
+
+	// Check that that sender node is operational
+	if node.Status() != river.NodeStatus_Operational {
+		return RiverError(Err_UNAUTHENTICATED, "Node is not operational").LogError(logger)
+	}
+
+	return nil
+}
+
+// CertGetter returns a GetClientCertFunc that provides a node-2-node client certificate.
+func CertGetter(logger *zap.SugaredLogger, wallet *crypto.Wallet) http_client.GetClientCertFunc {
+	const (
+		certTTL      = time.Hour * 24
+		certRenewGap = time.Minute
+	)
+
+	var (
+		lock sync.Mutex
+		cert *tls.Certificate
+		exp  time.Time
+	)
+
+	return func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		lock.Lock()
+		defer lock.Unlock()
+
+		if cert == nil || time.Now().Add(certRenewGap).After(exp) {
+			newCert, err := createCert(logger, wallet)
+			if err != nil {
+				return nil, err
+			}
+
+			cert = newCert
+			exp = time.Now().Add(certTTL)
+		}
+
+		return cert, nil
+	}
+}
+
+// createCert creates a node-2-node client certificate.
+// The certificate contains a custom extension with the node's address and a signature of the cert's public key.
+func createCert(logger *zap.SugaredLogger, wallet *crypto.Wallet) (*tls.Certificate, error) {
+	if wallet == nil {
+		return nil, RiverError(Err_INTERNAL, "No wallet provided").LogError(logger)
+	}
+
+	privateKeyA, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, AsRiverError(err, Err_INTERNAL).Message("Failed to generate private key").LogError(logger)
+	}
+
+	publicKeyDER, err := x509.MarshalPKIXPublicKey(&privateKeyA.PublicKey)
+	if err != nil {
+		return nil, AsRiverError(err, Err_INTERNAL).Message("Failed to marshal public key").LogError(logger)
+	}
+
+	signature, err := wallet.SignHash(crypto.TownsHashForCert.Hash(publicKeyDER))
+	if err != nil {
+		return nil, AsRiverError(err, Err_INTERNAL).Message("Failed to sign public key").LogError(logger)
+	}
+
+	extensionValue, err := asn1.Marshal(node2NodeCertExt{
+		Address:   wallet.Address.Hex(),
+		Signature: signature,
+	})
+	if err != nil {
+		return nil, AsRiverError(err, Err_INTERNAL).Message("Failed to marshal extra extension").LogError(logger)
+	}
+
+	template := &x509.Certificate{
+		SerialNumber:    big.NewInt(1),
+		Subject:         pkix.Name{CommonName: node2NodeCertName},
+		NotBefore:       time.Now(),
```

Comment:
```
template := &x509.Certificate{
		...
		NotBefore:       time.Now(),
}
```

Due to clock drift it is possible that the other nodes clock is slightly behind the clock of the node generating its client cert. So it is possible that when the certificate is used immediately after it is generated the other node might consider it not (yet) valid.

Maybe we should subtract 30s from `NotBefore` to compensate for small clock differences?

Reply:
Same issue for `NotAfter` if the remote nodes clock is just ahead of the local node.